File: src/utils/init_regions.py
import os
import sys
import pysam
import scipy as sp

def init_regions(fn_bams):
    # regions=init_regions(fn_bams)

    regions = []
    processed = []
    cnt = 0

    for i in range(len(fn_bams)):
        if not os.path.exists(fn_bams[i]):
            continue
        else:
            ### load bamfile
            IN = pysam.Samfile(fn_bams[i], 'rb')
            header_info = IN.header['SQ']
            strands = ['+', '-']
            for c in range(len(header_info)):
                if not (header_info[c]['SN'], header_info[c]['LN']) in processed:
                    for s in strands:
                        region = Region()
                        region.chr = header_info[c]['SN'])
                        region.chr_num = c
                        region.strand = s
                        region.start = 1
                        region.stop = header_info[c]['LN']
                        region.id = cnt
                        regions.append(region)
                        cnt += 1
                    processed.append((header_info[c]['SN'], header_info[c]['LN']))
            IN.close()
        break

    return sp.array(regions, dtype = 'object')

File: matlab/src/utils/init_regions.m
function regions=init_regions(fn_bams, organism)
% regions=init_regions(fn_bams, organism)

regions(1).chr = [];
regions(1).chr_num = [];
regions(1).strand = [];
regions(1).start = [];
regions(1).stop = [];
regions(1).offset =[];
regions(1).id = [];
regions(1).organism = [];

for i = 1:length(fn_bams),
    if exist(fn_bams{i}, 'file') == 0,
        continue;
    else
      regions = regions([]) ;
      header_info = get_header(fn_bams{i});
      strands = '+-';
      for c = 1:size(header_info, 1),
        %+strand
        for s = 1:length(strands),
          region = [];
          region.chr = header_info{c, 1};
          region.chr_num = c;
          region.strand = strands(s);
          region.start = 1;
          region.stop = header_info{c, 2}(1) ;
          switch strands(s)
           case '+',
            region.offset      = region.start - 1;
           case '-',
            region.offset      = region.stop + 1;
          end
          region.id = (c-1)*2+s;
          region.organism = [];
          if exist('organism', 'var')==1,
            region.organism = organism;
          end;
          regions(region.id) = region;
        end;
      end;
    end;
    break;
end;
